sdks/java/core/src/main/java/org/apache/beam/sdk/util/PaneInfoTracker.java

```diff
@@ -32,123 +32,206 @@
 
 import org.joda.time.Instant;
 
+import java.util.concurrent.ThreadLocalRandom;
+import javax.annotation.Nullable;
+
 /**
  * Determine the timing and other properties of a new pane for a given computation, key and window.
  * Incorporates any previous pane, whether the pane has been produced because an
  * on-time {@link AfterWatermark} trigger firing, and the relation between the element's timestamp
  * and the current output watermark.
  */
 public class PaneInfoTracker {
-  private TimerInternals timerInternals;
 
-  public PaneInfoTracker(TimerInternals timerInternals) {
-    this.timerInternals = timerInternals;
+  /**
+   * Interface for calculating and tracking pane info state.
+   */
+  public interface PaneInfoTrackerIface {
+    void clear(StateAccessor<?> state);
+
+    ReadableState<PaneInfo> getNextPaneInfo(
+        ReduceFn<?, ?, ?, ?>.Context context, final boolean isFinal);
+
+    void storeCurrentPaneInfo(ReduceFn<?, ?, ?, ?>.Context context, PaneInfo currentPane);
+
+    boolean hasState();
   }
 
   @VisibleForTesting
   static final StateTag<Object, ValueState<PaneInfo>> PANE_INFO_TAG =
       StateTags.makeSystemTagInternal(StateTags.value("pane", PaneInfoCoder.INSTANCE));
 
-  public void clear(StateAccessor<?> state) {
-    state.access(PANE_INFO_TAG).clear();
+
+  public static PaneInfoTrackerIface globalPaneInfoTracker() {
+    return new GlobalPaneInfoTracker();
+  }
+
+  public static PaneInfoTrackerIface localPaneInfoTracker(TimerInternals timerInternals) {
```

Comment:
The terms "global" and "local" are a bit vague. My understanding of the difference is that one is a specialized `PaneInfoTracker` for the case where EOW is unreachable, while the other is fully general.